internal/cloudfront/distribution.go

```diff
@@ -0,0 +1,149 @@
+	webACLID            string
+}
+
+// NewDistributionBuilder takes required arguments for a distribution and returns a DistributionBuilder
+func NewDistributionBuilder(defaultOriginDomain, description, priceClass, group string) DistributionBuilder {
+	return DistributionBuilder{
+		description:         description,
+		defaultOriginDomain: defaultOriginDomain,
+		priceClass:          priceClass,
+		group:               group,
+	}
+}
+
+// WithLogging takes in bucket address and file prefix to enable sending CF logs to S3
+func (b DistributionBuilder) WithLogging(bucketAddress, prefix string) DistributionBuilder {
+	b.logging = loggingConfig{
+		Enabled:       true,
+		BucketAddress: bucketAddress,
+		Prefix:        prefix,
+	}
+	return b
+}
+
+// WithTags takes in custom tags which should be present at the Distribution
+func (b DistributionBuilder) WithTags(tags map[string]string) DistributionBuilder {
+	b.tags = tags
+	return b
+}
+
+// WithTLS takes in an ACM certificate ARN and a Security Policy ID to enable TLS termination
+func (b DistributionBuilder) WithTLS(certARN, securityPolicyID string) DistributionBuilder {
+	b.tls = tlsConfig{
+		Enabled:          true,
+		CertARN:          certARN,
+		SecurityPolicyID: securityPolicyID,
+	}
+	return b
+}
+
+// WithIPv6 enables IPv6
+func (b DistributionBuilder) WithIPv6() DistributionBuilder {
+	b.ipv6Enabled = true
+	return b
+}
+
+// WithAlternateDomains takes a slice of domains to be configured as the Distribution's alternate domains
+func (b DistributionBuilder) WithAlternateDomains(domains []string) DistributionBuilder {
+	b.alternateDomains = domains
+	return b
+}
+
+// WithWebACL takes the ID of the Web ACL that should be associated with the Distribution
+func (b DistributionBuilder) WithWebACL(id string) DistributionBuilder {
+	b.webACLID = id
+	return b
+}
+
+// Build constructs a Distribution taking into account all configuration set by previous "With*" method calls
+func (b DistributionBuilder) Build() Distribution {
+	return Distribution{
+		DefaultOriginDomain: b.defaultOriginDomain,
+		Description:         b.description,
+		PriceClass:          b.priceClass,
+		Tags:                b.generateTags(),
+		Logging:             b.logging,
+		TLS:                 b.tls,
+		IPv6Enabled:         b.ipv6Enabled,
+		AlternateDomains:    b.alternateDomains,
+		WebACLID:            b.webACLID,
+	}
+}
+
+func (b DistributionBuilder) generateTags() map[string]string {
+	tags := b.defaultTags()
+	for k, v := range b.tags {
+		tags[k] = v
+	}
+	return tags
+}
+
+const (
+	managedByTagKey   = "cdn-origin-controller.gympass.com/owned"
+	managedByTagValue = "true"
+	groupTagKey       = "cdn-origin-controller.gympass.com/cdn.group"
+)
+
+func (b DistributionBuilder) defaultTags() map[string]string {
+	tags := make(map[string]string)
+	tags[managedByTagKey] = managedByTagValue
+	tags[groupTagKey] = b.group
+	return tags
+}
+
+// Distribution represents a CloudFront distribution
```

Comment:
What do you think in the next implementation the `Distribution` receives the `DefaultOrigin` and custom `Origin` as attributes? As Origin and Behavior have complex creation, the `DistributionBuilder` can receive `OriginBuilder` to delegate the way to build origins and behaviors.

The repository can receive only the `Distribution` and save all objects, I think makes sense in this case, because we every single time have to salve all `*DistributionConfig` on the AWS.

```suggestion
type Distribution struct {
...
DefaultOrigin Origin
CustomOrigin Origin
}
```
or

```suggestion
type Distribution struct {
...
Origins []Origin
}
```
Where `Origin` could have an attribute `type` to differentiate between `default` and `custom` (I think the first way is simpler).

In the repository:

Change the origin by distribution repo

```suggestion
type OriginRepository interface { -> type DistributionRepository interface {
```

```suggestion
func (r repository) Save(id string, d Distribution) error {
```